File: minesweeper/python/minesweeper.py
"""
Play the game of Minesweeper

V Buckley
05.20.2024
"""



import random

remainingMines = -1
settings = {
    "Mercy": True,
    "Easy fill": True,
    "Assisted start": False,
    "Mine move random": False,
}



def main():
    menuPrompt = "MINESWEEPER"
    menuOptions = ["Play", "Settings", "Quit"]
    menuOptionChoices = ["play", "settings", "quit"]
    
    loop = True

    while (loop):
        menuOption = get_string_selection(menuPrompt, menuOptions, 
            menuOptionChoices)

        if (menuOption == "play"):
            play_option()
        
        elif (menuOption == "settings"):
            settings_option()
        
        else:
            loop = False

    print()



def play_minesweeper(board, currentBoard, mines):
    """
    Allow the user to play minesweeper with the generated board
    """

    global remainingMines
    mode = "Reveal"
    firstClick = True

    remainingMines = mines

    won = False
    deathCoords = [-1, -1]

    while ((not won) and (deathCoords == [-1, -1])):
        print("\nMines Remaining: %d" % (remainingMines))
        print_board(currentBoard)
        
        print("\nCurrent mode: %s (ENTER to switch)" % (mode))

        xSelection = get_int_quit("X", 1, len(board))

        if (xSelection == -1):
            mode = switch_mode(mode)

        else:
            ySelection = get_int_quit("Y", 1, len(board[0]))

            if (ySelection == -1):
                mode = switch_mode(mode)
                
            else:
                realX = xSelection - 1
                realY = len(board[0]) - ySelection

                if (mode == "Flag"):
                    flag_mode(currentBoard, realX, realY)
                    
                else:
                    deathCoords = reveal_mode(board, currentBoard, realX, 
                        realY, firstClick)
                    firstClick = False
                    
                won = check_victory(board, currentBoard)
    
    return deathCoords


def play_option():
    """
    Menu option to play Minesweeper
    """
    
    difficultyPrompt = "Please select a difficulty"
    difficulties = ["Beginner", "Intermediate", "Expert", "Custom"]
    difficultyChoices = ["beginner", "intermediate", "expert", "custom"]

    difficulty = get_string_selection(difficultyPrompt, difficulties, 
        difficultyChoices)

    if (difficulty == "beginner"):
        width = 8
        height = 8
        mines = 10

    elif (difficulty == "intermediate"):
        width = 16
        height = 16
        mines = 40

    elif (difficulty == "expert"):
        width = 30
        height = 16
        mines = 99

    else:
        print()

        width = get_int("Width of the game board", 1, 99)
        height = get_int("Height of the game board", 1, 99)
        mines = get_int("Number of mines", 0, width * height)

    board = generate_game(width, height, mines)

    currentBoard = generate_blank_board(board)

    deathCoords = play_minesweeper(board, currentBoard, mines)
    
    finalBoard = final_board(board, currentBoard, deathCoords)

    if (deathCoords == [-1, -1]):
        print_board(currentBoard)
        print("\nCongratulations! You win!")
    
    else:
        print("\nSorry! You lose!")
    
    print_board(finalBoard)


def settings_option():
    """
    Allow the user to change the game settings
    """

    global settings

    loop = True

    while (loop):
        print("\nSettings")

        print("  -(1) Mercy: %s" % (settings["Mercy"]))
        print("  -(2) Easy fill: %s" % (settings["Easy fill"]))
        print("  -(3) Assisted start: %s" % (settings["Assisted start"]))
        print("  -(4) Mine move random: %s" % (settings["Mine move random"]))
        
        print("\n  -(5) Reset to defaults")
        print("  -(6) Back")

        print()
        setting = get_int("Choice", 1, 6)

        if (setting == 1):
            settings["Mercy"] = not settings["Mercy"]

        elif (setting == 2):
            settings["Easy fill"] = not settings["Easy fill"]

        elif (setting == 3):
            settings["Assisted start"] = not settings["Assisted start"]

        elif (setting == 4):
            settings["Mine move random"] = not settings["Mine move random"]
        
        elif (setting == 5):
            settings["Mercy"] = True
            settings["Easy fill"] = True
            settings["Assisted start"] = False
        
        else:
            loop = False


def get_string_selection(prompt, strings, stringChoices):
    """
    Get input from the user to select a string from a list
    """

    print("\n%s" % (prompt))

    validInput = False

    while (not validInput):
        for i in range(len(strings)):
            print("  -%s" % (strings[i]))

        userInput = input("\n").lower()

        if (userInput in stringChoices):
            validInput = True

        if (not validInput):
            print("\nPlease select a valid option")
    
    return userInput


def get_int(prompt, low, high):
    """
    Get user input of an integer between lowest and highest, inclusive
    """

    validInput = False

    while (not validInput):
        userInput = input("%s: " % (prompt))

        if (userInput.isdigit()):
            userInputInt = int(userInput)

            if (userInputInt >= low):
                if (userInputInt <= high):
                    validInput = True

        if (not validInput):
            print("Please enter an int between %d and %d" % (low, high))
    
    return userInputInt


def generate_game(width, height, mines):
    """
    Generate the Minesweeper game board
    """

    board = []

    for i in range(width):
        board.append([])

        for j in range(height):
            board[i].append(0)
    
    mineCounter = 0

    while (mineCounter < mines):
        mineX = random.randint(0, width - 1)
        mineY = random.randint(0, height - 1)

        if (board[mineX][mineY] != -1):
            board[mineX][mineY] = -1
            mineCounter += 1
    
    update_neighbor_counts(board)

    return board


def print_board(board):
    """
    Print the game board
    """

    widthNums = "      "
    for i in range(len(board)):
        widthNums += " %2d" % (i + 1)

    print(widthNums)

    horizontalBorder = "     +-"
    horizontalBorder += "---" * len(board)
    horizontalBorder += "-+"

    print(horizontalBorder)

    for i in range(len(board[0])):
        boardLine = " %2d  | " % (len(board[0]) - i)

        for j in range(len(board)):
            if (board[j][i] == -4):
                boardLine += "💥 "

            elif (board[j][i] == -3):
                boardLine += " - "
            
            elif (board[j][i] == -2):
                boardLine += " 🚩"
            
            elif (board[j][i] == -1):
                boardLine += "💣 "
            
            elif (board[j][i] == 0):
                boardLine += "   "

            else:
                boardLine += " " + str(board[j][i]) + " "

        boardLine += " |  %-2d" % (len(board[0]) - i)
        
        print(boardLine)
    
    print(horizontalBorder)
    print(widthNums)


def count_neighbors(board, x, y, cellType):
    """
    Count the number of neighboring cells of a given type to a given cell
    """
    
    counter = 0

    for i in range(-1, 2):
        if (((x + i) >= 0) and ((x + i) < len(board))):
            for j in range(-1, 2):
                if (((y + j) >= 0) and ((y + j) < len(board[0]))):
                    if board[x + i][y + j] == cellType:
                        counter += 1
    
    return counter


def generate_blank_board(board):
    """
    Generate the blank board to correspond to what the player has learned so 
        far
    """
    
    blankBoard = []

    for i in range(len(board)):
        blankBoard.append([])

        for j in range(len(board[0])):
            blankBoard[i].append(-3)
    
    return blankBoard
    

def get_int_quit(prompt, low, high):
    """
    Get user input of an integer between lowest and highest, inclusive
    Return -1 if nothing is entered to trigger a mode switch
    """
    
    validInput = False

    while (not validInput):
        userInput = input("%s: " % (prompt))

        if (userInput == ""):
            userInputInt = -1
            validInput = True
        
        else:
            if (userInput.isdigit()):
                userInputInt = int(userInput)

                if (userInputInt >= low):
                    if (userInputInt <= high):
                        validInput = True

        if (not validInput):
            print("Please enter an int between %d and %d" % (low, high))
    
    return userInputInt


def check_victory(board, currentBoard):
    """
    Check to see if the user has won
    """

    won = True

    for i in range(len(board)):
        for j in range(len(board[0])):
            if (board[i][j] != -1):
                if (board[i][j] != currentBoard[i][j]):
                    won = False

    return won


def flood_fill(board, currentBoard, x, y):
    """
    Reveal all of the cells that are 0 and their neighbors around the selected 
        cell
    """
    
    if ((x >= 0) and (x < len(board))):
        if ((y >= 0) and (y < len(board[0]))):
            if (currentBoard[x][y] == -3):
                currentBoard[x][y] = board[x][y]

                if (board[x][y] == 0):
                    flood_fill(board, currentBoard, x - 1, y - 1)
                    flood_fill(board, currentBoard, x - 1, y + 0)
                    flood_fill(board, currentBoard, x - 1, y + 1)
                    flood_fill(board, currentBoard, x + 0, y - 1)
                    flood_fill(board, currentBoard, x + 0, y + 1)
                    flood_fill(board, currentBoard, x + 1, y - 1)
                    flood_fill(board, currentBoard, x + 1, y + 0)
                    flood_fill(board, currentBoard, x + 1, y + 1)
    

def reveal(board, currentBoard, x, y, deathCoords, firstClick):
    """
    Reveal a cell
    """

    if (firstClick):
        if ((settings["Assisted start"]) and ((board[x][y]) != 0)):
            assisted_start(board, x, y)

        elif ((settings["Mercy"]) and ((board[x][y]) == -1)):
            mercy(board, x, y)
            

    if (board[x][y]) == -1:
        deathCoords = [x, y]
    
    else:
        flood_fill(board, currentBoard, x, y)
    
    return deathCoords


def reveal_around(board, currentBoard, x, y, deathCoords, firstClick):
    """
    Reveal all of the surrounding cells to a selected cell
    """
    
    for i in [-1, 0, 1]:
        if (((x + i) >= 0) and ((x + i) < len(board))):
            for j in [-1, 0, 1]:
                if (((y + j) >= 0) and ((y + j) < len(board[0]))):
                    if (currentBoard[x + i][y + j] == -3):
                        flags = count_neighbors(currentBoard, x, y, -2)

                        if (flags >= board[x][y]):
                            deathCoords = reveal(board, currentBoard, x + i,
                                y + j, deathCoords, firstClick)
    
    return deathCoords


def switch_mode(mode):
    """
    Switch gamemodes
    """
    
    if (mode == "Flag"):
        mode = "Reveal"

    else:
        mode = "Flag"

    return mode


def flag_mode(currentBoard, x, y):
    """
    Control what happens when a cell is selected while in flag mode
    """

    global remainingMines
    
    if (currentBoard[x][y] == -3):
        currentBoard[x][y] = -2
        remainingMines -= 1
    
    elif (currentBoard[x][y] == -2):
        currentBoard[x][y] = -3
        remainingMines += 1


def reveal_mode(board, currentBoard, x, y, firstClick):
    """
    Control what happens when a cell is selected while in reveal mode
    """

    global remainingMines

    deathCoords = [-1, -1]

    if (currentBoard[x][y] == -3):
        deathCoords = reveal(board, currentBoard, x, y, deathCoords, 
            firstClick)
    
    elif ((currentBoard[x][y] > 0) and (settings["Easy fill"])):
        deathCoords = reveal_around(board, currentBoard, x, y, deathCoords, 
            firstClick)   
    
    return deathCoords


def final_board(board, currentBoard, deathCoords):
    """
    Generate the final, ending board to be displayed if the user won or lost
    """

    finalBoard = generate_blank_board(board)

    if (deathCoords == [-1, -1]):
        for i in range(len(finalBoard)):
            for j in range(len(finalBoard[0])):
                if (currentBoard[i][j] == -3):
                    finalBoard[i][j] = board[i][j]

                elif (currentBoard[i][j] == -2):
                    finalBoard[i][j] = -2

                else:
                    finalBoard[i][j] = board[i][j]
            
    else:
        for i in range(len(finalBoard)):
            for j in range(len(finalBoard[0])):
                if (currentBoard[i][j] == -3):
                    finalBoard[i][j] = board[i][j]

                elif (currentBoard[i][j] == -2):
                    if (board[i][j] == -1):
                        finalBoard[i][j] = -2

                    else:
                        finalBoard[i][j] = board[i][j]

                else:
                    finalBoard[i][j] = board[i][j]

        finalBoard[deathCoords[0]][deathCoords[1]] = -4

    return finalBoard


def mercy(board, x, y):
    """
    Remove the bomb from the clicked location if it is the first click, if 
        possible
    """

    board[x][y] = 0

    if (settings["Mine move random"]):
        loop = True

        while (loop):
            mineX = random.randint(0, len(board) - 1)
            mineY = random.randint(0, len(board[0]) - 1)

            if ((mineX != x) or (mineY != y)):
                if (board[mineX][mineY] != -1):
                    board[mineX][mineY] = -1
                    loop = False
    
    else:
        loop = True
        i = 0
        j = 0

        while (loop):
            if (i != x) or (j != y):
                if (board[i][j] != -1):
                    loop = False

            if (loop == True):
                i += 1

                if (i >= (len(board))):
                    i = 0
                    j += 1

                    if (j >= (len(board[0]))):
                        i = x
                        j = y
                        loop = False

        board[i][j] = -1

    update_neighbor_counts(board)


def update_neighbor_counts(board):
    for l in range(len(board)):
        for m in range(len(board[0])):
            if (board[l][m] >= 0):
                neighbors = count_neighbors(board, l, m, -1)
                board[l][m] = neighbors


def assisted_start(board, x, y):
    """
    Assist the user so that their first click is always a zero (or best case 
        scenario if not possible)
    """

    k = 0
    l = 0
    
    for i in [-1, 0, 1]:
        if (((x + i) >= 0) and ((x + i) < len(board))):
            for j in [-1, 0, 1]:
                if (((y + j) >= 0) and ((y + j) < len(board[0]))):
                    if board[x + i][y + j] == -1:
                        board[x + i][y + j] = 0

                        if (settings["Mine move random"]):
                            loop = True

                            while (loop):
                                mineX = random.randint(0, len(board) - 1)
                                mineY = random.randint(0, len(board[0]) - 1)

                                if ((mineX < x - 1) or (mineX > x + 1)):
                                    if ((mineY < y - 1) or (mineY > y + 1)):
                                        if (board[mineX][mineY] != -1):
                                            board[mineX][mineY] = -1
                                            loop = False
                        
                        else:
                            loop = True

                            while (loop):
                                if (k not in [x - 1, x, x + 1]) or \
                                    (l not in [y - 1, y, y + 1]):
                                    if (board[k][l] != -1):
                                        loop = False
                                
                                if (loop == True):
                                    k += 1

                                    if (k >= (len(board))):
                                        k = 0
                                        l += 1

                                        if (l >= (len(board[0]))):
                                            k = x + i
                                            l = y + j
                                            loop = False

                            board[k][l] = -1

    if (board[x][y] == -1):
        mercy(board, x, y)

    update_neighbor_counts(board)



main()

File: minesweeper/c/minesweeper.c
/*
 * Play the game of Minesweeper
 * 
 * V Buckley
 * Started: 05.29.2024
 * 
 * v1.110
 */



/******************************* Header Files *******************************/

#include <stdio.h>
#include <stdlib.h>
#include <time.h>




/******************************* Definitions *******************************/

#define MAXLENGTH 1024




/***************************** Global Variables *****************************/

/*
 * All of the data needed to keep track of globally to run each game of Minesweeper
 */
struct minesweeperData {

    int width;
    int height;
    int mines;

    int *board;
    int *currentBoard;
};


/*
 * All of the editable settings options
 */
struct settingsData {

    int setSeed;

    int mercy;
    int easyFill;
    int assistedStart;
    int mineMoveRandom;
};




/*************************** Function Prototypes ***************************/

void generate_game(struct minesweeperData *data, struct settingsData *settings);
void play_minesweeper(struct minesweeperData *data);
void print_board(struct minesweeperData *data, int *board);
void generate_blank_board(struct minesweeperData *data);
void update_neighbor_counts(struct minesweeperData *data);
int count_neighbors(struct minesweeperData *data, int i, int j, int type);
int get_int_input(unsigned int low, unsigned int high);
int difficultyManager(struct minesweeperData *data);
int main_menu(struct minesweeperData *data, struct settingsData *settings);
void print_title();
int menu_selector();




/******************************** Functions ********************************/

int main() {

    int loop = 1, ret;
    struct minesweeperData data;
    struct settingsData settings;

    while (loop) {
        ret = main_menu(&data, &settings);
            
        play_minesweeper(&data);

        free(data.board);
        free(data.currentBoard);

        loop = 0;
    }

    return 0;
}



/*
 * Generate the Minesweeper game board
 * Args:
 *      data (minesweeperData *) - the struct containing all of the Minesweeper board data
 *      settings (settingsData *) - the editable settings that the user can control
 * Ret:
 *      None
 */
void generate_game(struct minesweeperData *data, struct settingsData *settings) {
    
    int i = 0, seed, mineCounter = 0, cellIdx;

    (data->board) = malloc(sizeof(int) * ((data->width) * (data->height)));

    if (!(data->board)) {
        perror("\nERROR: malloc failed");
        exit(1);
    }

    for (i; i < ((data->width) * (data->height)); i++) {
        (data->board)[i] = 0;
    }

    if ((settings->setSeed) == 1) {
        srand(time(NULL));

    } else {
        seed = get_int_input(0, 4294967295);
        srand(seed);
    }

    while (mineCounter < (data->mines)) {
        cellIdx = rand() % ((data->width) * (data->height));

        if ((data->board)[cellIdx] != 9) {
            (data->board)[cellIdx] = 9;
            mineCounter++;
        }
    }

    update_neighbor_counts(data);

}


/*
 * Allow the user to play minesweeper with the generated board
 * Args:
 *      data (minesweeperData *) - the struct containing all of the Minesweeper board data
 * Ret:
 *      None
 */
void play_minesweeper(struct minesweeperData *data) {

    int loop = 1, x, y;

    print_board(data, (data->board));

    while (loop) {
        print_board(data, (data->currentBoard));

        printf("\nReveal mode (enter 0 to switch modes)\n");

        printf("\nPlease select a column (x):\n");
        x = get_int_input(0, (data->width));

        if (!x) {
            printf("DEBUG: mode swap\n");

        } else {
            printf("\nPlease select a row (y):\n");
            y = get_int_input(0, (data->height));

            if (!y) {
                printf("DEBUG: mode swap\n");
            
            } else {
                printf("DEBUG: reveal\n");

                loop = 0;
            }
        }
    }

}


/*
 * Print the game board
 * Args:
 *      data (minesweeperData *) - the struct containing all of the Minesweeper board data
 * Ret:
 *      None
 */
void print_board(struct minesweeperData *data, int *board) {

    int i = 0, j, cellNum;

    printf("     ");
    for (i; i < (data->width); i++) {
        printf(" %-2d", i + 1);
    }
    printf("\n");

    printf("   +-");
    for (i = 0; i < (data->width); i++) {
        printf("---");
    }
    printf("-+\n");
    
    for (i = 0; i < (data->height); i++) {
        printf("%2d | ", ((data->height) - i));

        for (j = 0; j < (data->width); j++) {
            cellNum = board[j + (i * (data->width))];
            
            if ((cellNum >= 0) && (cellNum <= 8)) {     // Number (0-8)
                if (cellNum == 0) {         // Blank (8)
                    printf("\033[8m");
                
                } else if (cellNum == 1) {  // Blue (34)
                    printf("\033[34m");
                
                } else if (cellNum == 2) {  // Green (32)
                    printf("\033[32m");
                
                } else if (cellNum == 3) {  // Red (31)
                    printf("\033[31m");
                
                } else if (cellNum == 4) {  // Darkened Blue (2;34)
                    printf("\033[2;34m");
                
                } else if (cellNum == 5) {  // Darkened Red (2;31)
                    printf("\033[2;31m");
                
                } else if (cellNum == 6) {  // Cyan (36)
                    printf("\033[36m");
                
                } else if (cellNum == 7) {  // Darkened White (2)
                    printf("\033[2;90m");
                
                } else {                    // White (37)
                    printf("\033[2;37m");
                }

                printf("\033[1m %d \033[m", cellNum);
            
            } else if (cellNum == 9) {                  // Mine (9)
                printf("💣 ");

            } else if (cellNum == 10) {                 // Flag (10)
                printf(" 🚩");

            } else if (cellNum == 11) {                 // Hidden (11)
                printf("\033[2;90m - \033[m");

            } else if (cellNum == 12) {                 // Explosion (12)
                printf("💥 ");

            } else {
                fprintf(stderr, "\nERROR: Unexpected value in found in game board");
                exit(1);
            }
        }

        printf(" | %d\n", ((data->height) - i));
    }

    printf("   +-");
    for (i = 0; i < (data->width); i++) {
        printf("---");
    }
    printf("-+\n");

    printf("     ");
    for (i = 0; i < (data->width); i++) {
        printf(" %-2d", i + 1);
    }
    printf("\n");

}


/*
 * Generate the blank board to correspond to what the player has learned so far
 * Args:
 *      data (minesweeperData *) - the struct containing all of the Minesweeper board data
 * Ret:
 *      None
 */
void generate_blank_board(struct minesweeperData *data) {

    int i = 0;

    (data->currentBoard) = malloc(sizeof(int) * ((data->width) * (data->height)));

    if (!(data->currentBoard)) {
        perror("\nERROR: malloc failed");
        exit(1);
    }

    for (i; i < ((data->width) * (data->height)); i++) {
        (data->currentBoard)[i] = 11;
    }

}


/*
 * Count the number of mines neighboring each cell and update their values accordingly
 * Args:
 *      data (minesweeperData *) - the struct containing all of the Minesweeper board data
 * Ret:
 *      None
 */
void update_neighbor_counts(struct minesweeperData *data) {

    int i = 0, j, cellNum, neighbors;
    
    for (i; i < (data->height); i++) {
        for (j = 0; j < (data->width); j++) {
            cellNum = (data->board)[j + (i * (data->width))];

            if (cellNum <= 8) {
                neighbors = count_neighbors(data, i, j, 9);
                (data->board)[j + (i * (data->width))] = neighbors;
            }
        }
    }

}


/*
 * Count the number of neighboring cells of a given type to a given cell
 * Args:
 *      data (minesweeperData *) - the struct containing all of the Minesweeper board data
 *      i (int) - the row of the cell that is being checked
 *      j (int) - the column of the cell that is being checked
 *      type (int) - the type of cell we are searching for in the neighbors
 * Ret:
 *      counter (int) - the number of neighboring cells of the given type
 */
int count_neighbors(struct minesweeperData *data, int i, int j, int type) {

    int counter = 0, colOffset = -1, rowOffset;
    
    for (colOffset; colOffset <= 1; colOffset++) {
        if ((j + colOffset >= 0) && (j + colOffset < (data->width))) {
            for (rowOffset = -1; rowOffset <= 1; rowOffset++) {
                if ((i + rowOffset >= 0) && (i + rowOffset < (data->height))) {
                    if ((data->board)[(colOffset + j) + ((rowOffset + i) * (data->width))] == type) {
                        counter++;
                    }
                }
            }
        }
    }

    return counter;
}


/*
 * Get an integer input from the user between the given low and high values, inclusive
 * Args:
 *      low (int) - the smallest allowed int
 *      high (int) - the largest allowed int
 * Ret:
 *      num (int) - the selected int
 */
int get_int_input(unsigned int low, unsigned int high) {

    int validInput = 0;
    unsigned int num;
    char *ret;
    char userInput[MAXLENGTH];

    while(!validInput) {
        printf("Enter an int between %d and %u: ", low, high);
        fflush(stdout);

        ret = fgets(userInput, MAXLENGTH, stdin);
        if ((ferror(stdin)) || (!ret)) {
            perror("\nERROR: fgets failed");
            exit(1);
        }

        num = atoi(userInput);

        if ((num >= low) && (num <= high)) {
            validInput = 1;
        }
    }

    return num;
}


/*
 * Set all relevant data depending on the user selected difficulty 
 * Args:
 *      data (minesweeperData *) - the current game data struct
 * Ret:
 *      None
 */
int difficultyManager(struct minesweeperData *data) {
    
    int difficulty, ret = 0;

    printf("\nPlease select a difficulty:\n");
    printf("  0) Back\n");
    printf("  1) Beginner\n");
    printf("  2) Intermediate\n");
    printf("  3) Expert\n");
    printf("  4) Custom\n\n");

    difficulty = get_int_input(0, 4);

    if (difficulty == 0) {
        ret = 1;

    } else if (difficulty == 1) {
        (data->width) = 8;
        (data->height) = 8;
        (data->mines) = 10;

    } else if (difficulty == 2) {
        (data->width) = 16;
        (data->height) = 16;
        (data->mines) = 40;

    } else if (difficulty == 3) {
        (data->width) = 30;
        (data->height) = 16;
        (data->mines) = 99;

    } else if (difficulty == 4) {
        (data->width) = get_int_input(1, 99);
        (data->height) = get_int_input(1, 99);
        (data->mines) = get_int_input(0, ((data->width) * (data->height)));

    } else {
        fprintf(stderr, "\nERROR: invalid difficulty setting");
        exit(1);
    }

    return ret;
}


/*
 * Control the main menu of the game, allowing the user to change settings and begin playing
 * Args:
 *      data (minesweeperData *) - the current game data struct
 *      settings (settingsData *) - the editable settings that the user can control
 * Ret:
 *      ret (int) - TODO
 */
int main_menu(struct minesweeperData *data, struct settingsData *settings) {

    int selection, ret;

    print_title();

    selection = menu_selector();

    if (selection == 1) {
        ret = difficultyManager(data);

            if(ret == 0) {
                generate_game(data, settings);
                generate_blank_board(data);
            }
    }

    return ret;
}


/*
 * Print the title art and creator
 * Args:
 *      None
 * Ret:
 *      None
 */
void print_title() {
    
    printf("\n");
    printf(" ___________ _______ _______ _______ _______ ___   _   _ _______ _______ _______ _______ _______\n");
    printf("|    _   _  |_     _|    _  |    ___|    ___|   | | | | |    ___|    ___|    _  |    ___|    _  |\n");
    printf("|   | | | | | |   | |   | | |   |___|   |___|   | | | | |   |___|   |___|   |_| |   |___|   |_| |_\n");
    printf("|   | | | | | |   | |   | | |    ___|_____  |   | | | | |    ___|    ___|    ___|    ___|    ___  |\n");
    printf("|   | | | | |_|   |_|   | | |   |___ _____| |   |_| |_| |   |___|   |___|   |   |   |___|   |   | |\n");
    printf("|___| |_| |_|_______|___| |_|_______|_______|___________|_______|_______|___|   |_______|___|   |_|\n");
    printf("                                   Created by Grand Violet, 2024\n");

}


/*
 * Allow the user to select a menu option
 * Args:
 *      None
 * Ret:
 *      selection (int) - the chosen selections
 */
int menu_selector() {
    
    int selection;

    printf("  1) Play\n");
    printf("  2) Settings\n");
    printf("  0) Quit\n");

    selection = get_int_input(0, 2);

    return selection;
}
